core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -136,10 +138,25 @@ public void afterEntityAdded(EntityDescriptor<Solution_> entityDescriptor, Objec
         // beforeEntityAdded() has already added it to the notificationQueue
     }
 
+    public void beforeEntityRemoved(EntityDescriptor<Solution_> entityDescriptor, Object entity) {
+        Collection<EntityNotifiable<Solution_>> notifiables = notifiableRegistry.get(entityDescriptor);
+        if (!notifiables.isEmpty()) {
+            EntityNotification<Solution_> notification = Notification.entityRemoved(entity);
+            for (EntityNotifiable<Solution_> notifiable : notifiables) {
+                notifiable.addNotification(notification);
+            }
+        }
+        notificationQueuesAreEmpty = false;
```

Comment:
Should this really be outside the `if (!notifiables.isEmpty())` ?

Applies to every `before` method in this class.